Interface class

Flashing-Blinkenlights opened the pull request:
Only `interfaceUtils.py` has changed, the rest is from the `auto-documentation` branch.

nilsgawlik commented:
Hi, I merged the changes on dev back into this branch. 
I also made the example.py use the interface class so I could actually test it.
I also implemented batching to make sure it works, and without batching it wouldn't make sense to have this feature. I hope this didn't result in duplicate work!

Flashing-Blinkenlights commented:
> Hi, I merged the changes on dev back into this branch.
> I also made the example.py use the interface class so I could actually test it.
> I also implemented batching to make sure it works, and without batching it wouldn't make sense to have this feature. I hope this didn't result in duplicate work!

Not to worry, only cost me some nasty conflicts, but other then that a good idea. I used the opportunity to clean up the whole batching interface.
Switching between batching and non-batching is now safe and I adjusted the example.py to reflect that, also including basic instructions on how to use it.
I also added a fill command that I have thoroughly tested and works as anticipated, making the house generator less cluttered, and made the Interface call sendBlocks upon destruction, so even if you don't add any manual sendBlocks it will update. (Assuming a crash hasn't occurred)

nilsgawlik commented on interfaceUtils.py:
```diff
@@ -10,28 +10,34 @@
-        self.useBatching = useBatching
-        self.batchSize = batchSize
+        self.__buffering = False
+        self.bufferlimit = 4096
```
I think it's smart to set the default a little bit lower. I would suggest 1024 or 512. This is only slighly slower, but looks better in the game, and less blocks will be un-placed if, for example, the program pauses on an exception during development/debugging

nilsgawlik commented on interfaceUtils.py:
```diff
@@ -10,28 +10,34 @@
+    def __del__(self):
+        self.sendBlocks()
 
+    def requestBuildArea(self):
```
This is a change in behavior from the way getBuildArea worked before the interface class. In fact it returns less information, since it doesn't include the y-dimension.
The root problem is that the WorldSlice class takes a tuple, but the http interface returns a dictionary. This is just for historical reasons (WorldSlice is older that the build area functionality in the interface).
If that mismatch is to be fixed, I don't think changing the requestBuildArea functionality to the format of the WorldSlice class is the way to go.

Flashing-Blinkenlights replied:
I agree it's messy, what do you suggest?
Should we rework that part of WorldSlice?

nilsgawlik commented on interfaceUtils.py:
```diff
@@ -40,127 +46,139 @@ def requestBuildArea(self):
 
     def getBlock(self, x, y, z):
-        """**Returns the name of a block in the world.**"""
+        """**Return the name of a block in the world**."""
```
```suggestion
        """**Return the name of a block in the world.**"""
```

nilsgawlik commented on interfaceUtils.py:
```diff
@@ -40,127 +46,139 @@ def requestBuildArea(self):
-    def _setBlockDirect(self, x, y, z, blockStr):
-        """**Places a block in the world.**"""
+    def placeBlock(self, x, y, z, str):
+        """**Place a single block in the world**."""
```
```suggestion
        """**Place a single block in the world.**"""
```

nilsgawlik commented on interfaceUtils.py:
```diff
@@ -40,127 +46,139 @@ def requestBuildArea(self):
 
-    def _setBlockDirect(self, x, y, z, blockStr):
-        """**Places a block in the world.**"""
+    def placeBlock(self, x, y, z, str):
```
I renamed this to blockStr intentionally, because str is an internal function

Flashing-Blinkenlights replied:
Fair point, will amend.

nilsgawlik commented on interfaceUtils.py:
```diff
@@ -40,127 +46,139 @@ def requestBuildArea(self):
 
-    def _placeBlockBatched(self, x, y, z, blockStr, limit=50):
-        """**Place a block in the buffer and send if the limit is exceeded.**"""
+    def toggleBuffer(self):
```
This function seems unnecessary, why would anybody use it instead of just setting the buffer to what they want it to be directly?

Flashing-Blinkenlights replied:
For convenience and testing's sake, but we can get rid of it.

nilsgawlik replied:
Ok 👍

nilsgawlik commented on interfaceUtils.py:
```diff
@@ -40,127 +46,139 @@ def requestBuildArea(self):
+        return self.Buffering
+
+    @property
+    def Buffering(self):
```
I think the property should be lower-case. I know that in languages like C# uppercase is convention, but I think in python no such convention exists, and personally I prefer lowercase, too. I don't think the user needs to know that this is a property and not a field.

nilsgawlik commented on interfaceUtils.py:
```diff
@@ -40,127 +46,139 @@ def requestBuildArea(self):
+            self.sendBlocks()
+            print("Buffering has been deactivated.")
+
+    def placeBlockBatched(self, x, y, z, str, limit=50):
```
Like with placeBlock, `str` should be kept as `blockStr` so it doesn't override the `str()` function.

nilsgawlik commented on interfaceUtils.py:
```diff
@@ -40,127 +46,139 @@ def requestBuildArea(self):
-
         url = 'http://localhost:9000/blocks?x={}&y={}&z={}'.format(x, y, z)
-        body = str.join("\n", ['~{} ~{} ~{} {}'.format(*bp)
+        body = str.join("\n", ['{} {} {} {}'.format(*bp)
```
The relative coordinates were intentional. Leaving the `~` out breaks the relative positioning functionality of sendBlocks, and makes the optional x,y,z parameters useless.

nilsgawlik reviewed (commented):
Hi! 
sorry if there seem to be a lot of comments, it feels like some of these might just be unintentioal reverted changes because something was not merged properly.

There are two more general things I want to point out:

1) I think we need a unified word for buffering vs. batching. The http interface uses the word 'batching'. If we use buffering for the Interface class it'll be confusing, so my suggestion would be to change all the occurences of 'buffer'/'buffering' to 'batch'/'batching'

2) I see that all doc comments are like `**Some text**.` with the dot after the highlight-markers. Why would the dot not be included in the highlighted text? I thought this was a mistake, but it seems intentional.

Flashing-Blinkenlights commented on example.py:
```diff
@@ -44,63 +54,60 @@ def heightAt(x, z):
             for z in range(z1 + 1, z2 - 1):
-                interface.setBlock(x, y, z, "air")
+                # check what's at that place and only delete if not air
+                if "air" not in interface.getBlock(x, y, z):
```
I was thinking to use it as an example for getBlock and setBlock since they don't turn up in the rest of the code, but I don't mind reverting it.

nilsgawlik replied:
Ah ok, makes sense. Hm I don't think this is the way to do it, since I don't think it's a good example for using getBlock, but it would be cool if there was something that uses getBlock in the example.

Flashing-Blinkenlights replied:
Agreed. I'll leave it like that but we can rework it when reworking the example/s.

Flashing-Blinkenlights commented on example.py:
```diff
@@ -10,32 +10,42 @@
 It is not meant to be imported.
 """
 __all__ = []
+__author__ = "Nils Gawlik <nilsgawlik@gmx.de>"
```
I think that has something to do with the dev merge and me overseeing it (thinking that that change removal was done on another branch and would be cleared during that merge, but I wasn't thinking straight).

Flashing-Blinkenlights commented:
1. I found buffering to make more sense, since the place where the blocks are buffered can be called the buffer.
When using batch (as I started off doing), I realised that I found buffer more convenient and transparent since buffering is a common practice, whereas 'batch' is more used for a repetitive operation on multiple objects.
Personally, I would prefer to stick with 'buffer', but I also respect that it is sensible to use the same terminology as the interface.
I leave the choice up to you
2. According to PEP 257 "[In a docstring the] First line should end with a period", which means the `**` needs to come before the `.`, otherwise it risks confusing any machine trying to read the docstring.
Question is, should we just do it the way it makes sense to us or be a stickler and conform to the guideline?

Flashing-Blinkenlights commented on interfaceUtils.py:
```diff
@@ -10,28 +10,34 @@
         self.offset = offset
-        self.useBatching = useBatching
-        self.batchSize = batchSize
+        self.__buffering = False
```
To prevent people from accessing it directly, since that might lead to problems

nilsgawlik commented:
1. I think I agree that buffer is the better word. And I also noticed that in the documentation for the http interface I actually call it 'placing blocks in bulk', not nothing about batches there either.
2. Ah ok, makes sense 👍
